Notify WebDriver BiDi about "committed" navigations

OrKoN opened the pull request:
This PR adds integration with the WebDriver BiDi spec to notify automation sessions when a navigation results in a new document being loaded.  See https://github.com/w3c/webdriver-bidi/issues/788 and https://github.com/w3c/webdriver-bidi/pull/855.

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.

When you submit this PR, and each time you edit this comment (including checking a checkbox through the UI!), PR Preview will run and update it. As such make any edits in one go and only after PR Preview has run.

If you think your PR is ready to land, please double-check that the build is passing and the checklist is complete before pinging.
-->

- [x] At least two implementers are interested (and none opposed):
   * Chrome
   * Firefox
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * Tests will be written as part of the WebDriver BiDi spec (changes not observable otherwise).
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://github.com/GoogleChromeLabs/chromium-bidi/issues/2982
   * Gecko: https://github.com/w3c/webdriver-bidi/issues/788
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=281932
   * Deno (only for timers, structured clone, base64 utils, channel messaging, module resolution, web workers, and web storage): …
   * Node.js (only for timers, structured clone, base64 utils, channel messaging, and module resolution): …
- [x] Corresponding [HTML AAM](https://w3c.github.io/html-aam/) & [ARIA in HTML](https://w3c.github.io/html-aria/) issues & PRs: N/A
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: N/A
- [x] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/10910/browsing-the-web.html" title="Last updated on Jan 31, 2025, 5:14 AM UTC (90a0681)">/browsing-the-web.html</a>  ( <a href="https://whatpr.org/html/10910/ad1450d...90a0681/browsing-the-web.html" title="Last updated on Jan 31, 2025, 5:14 AM UTC (90a0681)">diff</a> )
<a href="https://whatpr.org/html/10910/infrastructure.html" title="Last updated on Jan 31, 2025, 5:14 AM UTC (90a0681)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/10910/ad1450d...90a0681/infrastructure.html" title="Last updated on Jan 31, 2025, 5:14 AM UTC (90a0681)">diff</a> )

sadym-chromium reviewed (commented):
LGTM

jgraham commented:
So in 6.4.1 in _attempt to populate the history entry's document_ (covering the case of navigating to a document without a DOM) seems to run a the navigation failed steps (and so emits a navigation failed event), but the document isn't null in that case so we'll still send the navigation committed event in 7.4. Possibly we need to move emitting this event up into step 6?

OrKoN commented:
> So in 6.4.1 in _attempt to populate the history entry's document_ (covering the case of navigating to a document without a DOM) seems to run a the navigation failed steps (and so emits a navigation failed event), but the document isn't null in that case so we'll still send the navigation committed event in 7.4. Possibly we need to move emitting this event up into step 6?

Thanks for spotting this. On one hand there is a new document committed by the navigation (that might be relevant for the client), on the other hand the events get confusing (failed, then committed) and IIUC this is about error pages where the only useful action is to reload. I guess moving to 6 would be an option or checking saveExtraDocumentState or document's [salvageable](https://whatpr.org/html/10910/document-lifecycle.html#concept-document-salvageable) state.

OrKoN commented:
Moved to step 6.

OrKoN commented:
@domenic could you please take a look when you have a chance?

domenic commented on source:
```diff
@@ -102097,17 +102099,33 @@ location.href = '#foo';</code></pre>
+        data-x="navigation-status-url">url</span> is <var>document</var>'s <span
+        data-x="concept-document-url">URL</span> if <var>entry</var>'s <span
+        data-x="she-document-state">document state</span>'s <span
+        data-x="document-state-document">document</span> is not null.</p></li>
```
Put the if at the beginning of the line.

domenic reviewed (commented):
I'm a little worried about this running before the history state is updated. In particular, the active history entry etc. will still be on the old document. Is that OK, in the WebDriver BiDi context?

OrKoN commented:
> I'm a little worried about this running before the history state is updated. In particular, the active history entry etc. will still be on the old document. Is that OK, in the WebDriver BiDi context?

WebDriver BiDi spec does depend on the history state for this event. In practice, I believe the history state would be completed before the event can be handled by the client. I think it would be great to emit when the state defined in the HTML spec is in the consistent state.

Is the history state usually updated in the completion steps of this algorithm? Should we move the event to after the completion steps are run? What would you recommend?

domenic commented:
History state is updated in the monster algorithm [apply the history step](https://html.spec.whatwg.org/#apply-the-history-step). Note that that algorithm is called for all history step updates, including traverses and reloads, not just the push/replace navigations you're interested in.

The algorithm you're modifying, "attempt to populate the history entry's document", is called in step 12.8.7 of "apply the history step". It basically does the fetch of the new document, but there are many steps left until we get to what I think of as the actual commit. The completion steps for it don't fully finish updating the history step: instead they enqueue an update which runs later, in step 14. Step 14.10 or 14.11 unloads the previous document, and then after that process is done it does a callback to _afterPotentialUnloads_.

_afterPotentialUnloads_ (declared in step 14.12) is where things get interesting. By its step 2, the "browser process" is all updated. But the document itself is not yet updated to realize that it's active, necessarily: e.g., its navigation API state is not yet initialized, and same-document navigations or bfcache traversals haven't updated the relevant JS APIs. That happens in [update document for history step application](https://html.spec.whatwg.org/#update-document-for-history-step-application).

So I suspect what you want is to insert something either after step 2 of _afterPotentialUnloads_, if you only care about the "browser process", or at the end of "update document for history step application", if you care about the document's state. The latter sounds safer.

The hardest part will be filtering so that you only do this on cross-document push/replace navigations, but I think the "update document for history step application" algorithm has enough local state that you should be able to do this. (_navigationType_ and _documentIsNew_ should do it, I think.)

OrKoN commented:
thanks @domenic It sounds like WebDriver BiDi would be interested in all situations when documentIsNew. I have updated the PR and moved the call to "update document for history step application" before any document scripts may run. 

@jgraham could you please take a look at the proposal as well?

domenic commented on source:
```diff
@@ -104297,6 +104299,16 @@ location.href = '#foo';</code></pre>
     <p>If <var>documentIsNew</var> is true, then:
 
     <ol>
+      <li><p>If <var>document</var>'s <span data-x="concept-document-navigation-id">during-loading
```
I think you can assert it is not null instead. At least, I cannot see how it would ever become null by this point.

domenic commented on source:
```diff
@@ -104297,6 +104299,16 @@ location.href = '#foo';</code></pre>
     <p>If <var>documentIsNew</var> is true, then:
```
Since you are not filtering for push/replace navigations, this will also trigger for back/forward/reload navigations. Is that intended?
